### ISSUE TITLE
RenderProps configuration

### ISSUE THREAD

**Issue by webberwang** [codelab-ai/codelab.ai#84]
Each prop can be 1 of 2 types of renderProps

This should pass 1 level
```ts
    const renderProps = {
      data: {
        renderProps: true,
        value: 'pass 1 level',
      },
    }
```

This should pass to leaf level
```ts
    const renderProps = {
      data: {
        renderProps: 'leaf',
        value: 'pass to leaf',
      },
    }
```

Write unit tests for both, the 1st one might already have one? For 1st unit tests, add a conditional that grandchildren should not receive the prop

----------------------

We only require `filterRenderProps`, which knows how to filter both types of renderProps, we then merge them into `props`. Since we can identify renderProps by their config, we can just use `props` from now

Basically it's just a way to merge parent props into child props based on some criteria, we wouldn't need an extra class property `leafRenderProps`

---------------------

Now `rootRenderProps` can be mapped via a `convertToLeafRenderProps` then just passed as regular props without the distinction of a rootRenderProps!

We do the conversion as early as possible, so inside TreeDom's return component function

We are essentially building a renderProps API, then we use it with rootRenderProps

**Comment by VienNguyen24hdev** [comment]
via Hubstaff
User: Nguyen Luong Vien

Project: Codelab - https://app.hubstaff.com/projects/1093588
Date Range: 09/09/20 - 09/10/20
Work session total: 2:41:22
Billable: Yes

Grand total: 2:41:22
